### PR TITLE
fix: support sparse B in Agent FPI inference

### PR DESCRIPTION
**Pull request by docxology**
## Summary

This is a reduced replacement for the sparse-B portion of PR #443, containing only the targeted runtime fixes requested in review.

- Preserve unbatched JAX `BCOO` transition matrices at the `Agent` constructor boundary.
- Validate sparse tensors without calling dense-only `jnp.sum`.
- Densify the one-dimensional predictive belief returned by `compute_expected_state` so downstream inference receives the documented dense belief representation.
- Add one focused Agent-vs-dense regression test.

No documentation sidecars, generated Sphinx artifacts, notebooks, examples, benchmarks, or unrelated repository files are included.

## Verification

- `python3 -m py_compile pymdp/agent.py pymdp/control.py pymdp/utils.py test/test_jax_sparse_backend.py`
- `uv run ruff check pymdp/agent.py pymdp/control.py pymdp/utils.py test/test_jax_sparse_backend.py`
- `uv run --with pytest --with pytest-cov --with pytest-xdist python -m pytest -q test/test_jax_sparse_backend.py --no-cov`
  - 3 passed
- `git diff --check`

### CODE REVIEW

**Review by conorheins — CHANGES_REQUESTED** [review]
This is a good start to introducing sparse tensors for FPI inference, but i think we shouldn't allow sparse tensors to make their way into the Agent class until we can use them for all agent methods, not just infer_states (so infer_policies, paramter learning, etc.). Here's a list of places that Codex spotted incompatibility

### Findings

1. **P1 — Sparse `B` still breaks normal policy inference and JIT.**  
   Densification happens only after `factor_dot`, but the sparse contraction itself fails when `infer_policies()` traces it through `lax.scan`. A minimal identity-BCOO Agent raises `TracerBoolConversionError` at [[control.py:402–403](https://github.com/infer-actively/pymdp/blob/e6bd3013e95731a5d9a34a09345db36c1e24d976/pymdp/control.py#L402-L403)](https://github.com/infer-actively/pymdp/blob/e6bd3013e95731a5d9a34a09345db36c1e24d976/pymdp/control.py#L402-L403), originating from traced Python logic in [[maths.py:164–167](https://github.com/infer-actively/pymdp/blob/e6bd3013e95731a5d9a34a09345db36c1e24d976/pymdp/maths.py#L164-L167)](https://github.com/infer-actively/pymdp/blob/e6bd3013e95731a5d9a34a09345db36c1e24d976/pymdp/maths.py#L164-L167). Jitted `update_empirical_prior()` fails similarly. This blocks the standard `infer_states → infer_policies` Agent cycle.

2. **P2 — Explicit sparse batching is accepted but not consistently usable.**  
   A natural `BCOO.fromdense()` over `(batch, states, states, actions)` has `n_batch=0`. The constructor accepts it, but later `vmap` fails because it requires a BCOO batch dimension. A singleton sparse batch also reaches dense-only `jnp.broadcast_to` and raises `TypeError` at [[agent.py:252–273](https://github.com/infer-actively/pymdp/blob/e6bd3013e95731a5d9a34a09345db36c1e24d976/pymdp/agent.py#L252-L273)](https://github.com/infer-actively/pymdp/blob/e6bd3013e95731a5d9a34a09345db36c1e24d976/pymdp/agent.py#L252-L273).

3. **P2 — `learn_B=True` fails with sparse `B`.**  
   The update branch returns dense `E_qB`, while the skip branch returns sparse `self.B`; `lax.cond` rejects the different PyTree structures at [[agent.py:609–622](https://github.com/infer-actively/pymdp/blob/e6bd3013e95731a5d9a34a09345db36c1e24d976/pymdp/agent.py#L609-L622)](https://github.com/infer-actively/pymdp/blob/e6bd3013e95731a5d9a34a09345db36c1e24d976/pymdp/agent.py#L609-L622).

4. **P2 — Some valid sparse normalization calls crash.**  
   When reduction produces a dense scalar, the unconditional `.todense()` at [[utils.py:117–118](https://github.com/infer-actively/pymdp/blob/e6bd3013e95731a5d9a34a09345db36c1e24d976/pymdp/utils.py#L117-L118)](https://github.com/infer-actively/pymdp/blob/e6bd3013e95731a5d9a34a09345db36c1e24d976/pymdp/utils.py#L117-L118) raises `AttributeError`. Reproduction: `validate_normalization(BCOO([0.5, 0.5]), axis=0)`.

A narrower additional gap is `B_action_dependencies=[[]]`: sparse `B` reaches dense-only `jnp.expand_dims` at [[agent.py:1115](https://github.com/infer-actively/pymdp/blob/e6bd3013e95731a5d9a34a09345db36c1e24d976/pymdp/agent.py#L1115)](https://github.com/infer-actively/pymdp/blob/e6bd3013e95731a5d9a34a09345db36c1e24d976/pymdp/agent.py#L1115).

I think given all this, we should this PR should be one feature into a more comprehensive fix that introduces sparse BCOO B tensors into all the places where the B tensor get used. It expands the scope of the fix but since this is a fundamental extension to the types of B tensors stored on the `Agent` class, I think it should make sure to not break all the things that get touched by `B`

**Comment by docxology** [comment]
Thanks @conorheins. I've expanded the implementation to address all reported findings across the complete Agent lifecycle when using sparse `BCOO` transition tensors:

1. **Policy inference & JIT traceability (`pymdp/maths.py`, `pymdp/control.py`)**:
   - Replaced untraced boolean masking in `spm_dot_sparse` with static dimension resolution and pure-JAX `sparse_sum`, eliminating `TracerBoolConversionError` during `lax.scan` tracing in `infer_policies()`.
   - Verified that `infer_policies()` and `update_empirical_prior()` match dense calculations exactly and compile cleanly under `jax.jit`.

2. **`learn_B` / `infer_parameters` structure consistency (`pymdp/agent.py`)**:
   - Ensured PyTree return types in `Agent.infer_parameters` match across both branches of `lax.cond` by consistently densifying skipped transition models, preventing PyTree type mismatches.

3. **Normalization validation on scalar reductions (`pymdp/utils.py`)**:
   - Updated `validate_normalization` to check for `.todense()` presence, safely handling 1D/scalar reductions without raising `AttributeError`.

4. **Empty action dependencies `B_action_dependencies=[[]]` (`pymdp/agent.py`)**:
   - Added sparse `BCOO` index-padding support in `_flatten_B_action_dims` for trivial action dimensions instead of calling dense `jnp.expand_dims`.

5. **End-to-end regression tests (`test/test_jax_sparse_backend.py`)**:
   - Added `test_sparse_b_agent_full_cycle` covering the complete `infer_states → update_empirical_prior → infer_policies → sample_action` cycle under JIT with direct comparisons against dense agents.

All 3 targeted sparse test suites pass cleanly.
